opal/mca/common/ucx/common_ucx.c

```diff
@@ -138,18 +138,30 @@ static void opal_common_ucx_mca_fence_complete_cb(int status, void *fenced)
 void opal_common_ucx_mca_proc_added(void)
 {
 #if HAVE_DECL_UCM_TEST_EVENTS
-    static int warned = 0;
-    static char *mem_hooks_suggestion = "Pls try adding --mca opal_common_ucx_opal_mem_hooks 1 "
-                                        "to mpirun/oshrun command line to resolve this issue.";
+    static int warned      = 0;
+    const char *suggestion;
     ucs_status_t status;
 
     if (!warned) {
-        status = ucm_test_events(UCM_EVENT_VM_UNMAPPED);
+        if (opal_common_ucx.opal_mem_hooks) {
+            suggestion = "Please check OPAL memory events infrastructure.";
+#if HAVE_DECL_UCM_TEST_EXTERNAL_EVENTS
+            status     = ucm_test_external_events(UCM_EVENT_VM_UNMAPPED);
+#endif
+        } else {
+            suggestion = "Pls try adding --mca opal_common_ucx_opal_mem_hooks 1 "
+                         "to mpirun/oshrun command line to resolve this issue.";
+#if !HAVE_DECL_UCM_TEST_EXTERNAL_EVENTS
+        }
+        {
+#endif
+            status     = ucm_test_events(UCM_EVENT_VM_UNMAPPED);
+        }
+
```

Comment:
code still seems too complicated,
1. move mem hooks test to separate function
2. try to simplify the combination of if/ifdef

Reply:
ok, done